Mocking EventHub Publishing Samples

ACHYUT001 opened the pull request:
@jsquire ,

I've started out writing the mocks.
Here are a few for Publishing, requesting your thoughts.
Should I cover all the scenarios as done in the [publishing sample](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample04_PublishingEvents.md)? 

I also needed your thoughts in understanding the points to cover in the markdown file of the sample.

Thanks

Commit:
Mocking EventHub Publishing Samples

jsquire commented:
Hi @ACHYUT001.   Thanks very much for helping to improve the Azure SDK experience!

> Should I cover all the scenarios as done in the publishing sample?

I don't think that we need to go over each of the sample scenarios.   I see the goal for the mocking samples as demonstrating the high-level concept for the major scenarios.   For example, the most important would be:

- How do I mock publishing a batch of events?   
  - Demonstrating how to use `Moq` to setup the `EventHubProducerClient` to handle `SendAsync`
  - Demonstrating how to use the `EventHubsModelFactory` to create an `EventDataBatch` and use the callback to pass a result to `TryAdd`

- How do I mock reading events from the `EventHubConsumer`?
  - Demonstrate how to use the `EventHubsModelFactory to create an `EventData` instance with the broker-owned properties set.
  - Demonstrate how to use `Moq` to setup the enumerable for `ReadEventsFromPartitionAsync`

- How do I mock reading events from the `PartitionReceiver`?
  - Demonstrate how to use the `EventHubsModelFactory to create an `EventData` instance with the broker-owned properties set.
  - Demonstrate how to use `Moq` to setup the enumerable for `ReceiveEventsAsync`

- How do I mock invoking my event and error handlers for the `EventProcessorClient`?  _(separate sample in the processor package)_
  - Demonstrate using the `EventHubsModelFactory` to create the types needed by the event args  

The issue outlines some of the other types around Event Hubs metadata, which we can demonstrate using the producer or consumer.

Generally speaking, we just want to help people realize that they don't need an interface to enable mocking - the virtual members can be mocked the same way.   We also want to help them find the `EventHubsModelFactory`, as it's critical to the mocking story and oft overlooked.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/tests/Snippets/Sample09_MockingEventHubPublisher.cs:
```diff
@@ -0,0 +1,205 @@
+            #region Snippet:EventHubs_Sample09_EventBatchMock
+
+            var size = 1024;
+            var store = new List<EventData> { new EventData(Array.Empty<byte>()), new EventData(Array.Empty<byte>()) };
```
I wouldn't focus on pre-populating the batch via the store;  its more intended to let developers peek into it and confirm what they've added to the batch.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/tests/Snippets/Sample09_MockingEventHubPublisher.cs:
```diff
@@ -0,0 +1,205 @@
+            var size = 1024;
+            var store = new List<EventData> { new EventData(Array.Empty<byte>()), new EventData(Array.Empty<byte>()) };
+            var mockResponse = new Mock<Response>();
+            Response<EventDataBatch> response = Response.FromValue(EventHubsModelFactory.EventDataBatch(size, store), mockResponse.Object);
```
The `Response` type is meant to represent  an HTTP response.   The Event Hubs clients don't use this, as they're AMQP-based.    I'm surprised this compiles, given that the return type of `CreateBatchAsync` is `ValueTask<EventDataBatch>`

What we'll want here is something more like:

```csharp
var mockEventBatch = EventHubsModelFactory.EventDataBatch(mockBatchSizeInBytes, eventStore);

mockEventHubProducerClient
    .Setup(c => c.CreateBatchAsync(It.IsAny<CancellationToken>()))
    .Returns(new ValueTask(mockEventBatch));
```

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/tests/Snippets/Sample09_MockingEventHubPublisher.cs:
```diff
@@ -0,0 +1,205 @@
+                var eventBody = new BinaryData("This is an event body");
+                var eventData = new EventData(eventBody);
+
+                if (!eventBatch.TryAdd(eventData))
```
There's a [`tryAddCallback`](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubsModelFactory.cs#L126) parameter that let's the result of `TryAdd` be mocked; it would probably be a good idea to show this as well.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/tests/Snippets/Sample09_MockingEventHubPublisher.cs:
```diff
@@ -0,0 +1,205 @@
+                await producer.CloseAsync();
+            }
+
+            #endregion
```
We probably want to show validation like developers would be doing in a test.   Creating a set of events up front, adding them to the batch, and then verifying that each event is in the batch would be helpful.   Something like:

```csharp

var expectedEvents = GenerateEventList( ... );

// Create the producer and batch the way we're doing now...

// Let's ensure that we see the Send call.

mockEventHubProducerClient
    .Setup(c => c.SendAsync(mockEventBatch, It.IsAny<CancellationToken>())
    .Returns(Task.CompletedTask)
    .Verifiable();

foreach (var eventData in expectedEvents)
{
     if (!eventBatch.TryAdd(eventData))
     {
          // My callback rejected the event....
     }
}

await producer.SendAsync(eventBatch, CancellationToken.None);

// Did the events that I think that I added to the batch actually make it into the
// batch?

for (var index = 0; index < expectedEvents.Cout; ++index)
{
    Assert.AreEqual(expectedEvents[index].EventBody.ToString(), eventStore[index].EventBody.ToString());
}

mockEventHubProducerClient.VerifyAll();
```

jsquire replied:
I'm deliberately using the `Assert.AreEqual` form here, so that it is more easily translatable for developers using xUnit or MSTest or another test framework.

Commit:
added mock for tryadd using callback
added assert statement on expected values

Commit:
added logic for mocking PartitionReceiver, EventProcessorClient
added demos to create event hub metadata types

ACHYUT001 commented on sdk/eventhub/Azure.Messaging.EventHubs/tests/Snippets/Sample10_MockingReadingEventsTests.cs:
```diff
@@ -0,0 +1,254 @@
+        }
+
+        private IAsyncEnumerable<PartitionEvent> GetMockEventsFromPartition(string partitionId,LastEnqueuedEventProperties lastEnqueuedEventProperties = default)
+        {
```
Hi @jsquire,
I need your input here, what is the best way to create the IAsyncEnumerable response for mocking ReadEventsFromPartitionAsync.

ACHYUT001 replied:
When I'm using this method here, I would need to make this method "async" but then I dont have an "await" statement in the method...😅
So I was wondering is there an alternative to do this
Thanks

ACHYUT001 commented on sdk/eventhub/Azure.Messaging.EventHubs/tests/Snippets/Sample10_MockingReadingEventsTests.cs:
```diff
@@ -0,0 +1,254 @@
+                {
+                    LastEnqueuedEventProperties properties =
+                        partitionEvent.Partition.ReadLastEnqueuedEventProperties();
+
```
@jsquire,
Should we assert the values here?

ACHYUT001 commented on sdk/eventhub/Azure.Messaging.EventHubs/tests/Snippets/Sample10_MockingReadingEventsTests.cs:
```diff
@@ -0,0 +1,254 @@
+            await using var scope = await EventHubScope.CreateAsync(1);
+
+            #region Snippet:EventHubs_Sample10_MockingReadPartitionTrackLastEnqueued
+
```
@jsquire ,
In the lines 149 -180, I've demonstrated how to use EventHubsModelFactory to create Event Hub meta data types and also mocked them.
Is this okay?
Or should we look at breaking them down to individuals?

ACHYUT001 commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Snippets/Sample07_MockingEventProcessorClientTest.cs:
```diff
@@ -0,0 +1,139 @@
+            mockEventProcessorClient
+                .Setup(c => c.StopProcessingAsync(It.IsAny<CancellationToken>()))
+                .Returns(Task.CompletedTask);
+
```
@jsquire ,

> How do I mock invoking my event and error handlers for the EventProcessorClient? (separate sample in the processor package)

Demonstrate using the EventHubsModelFactory to create the types needed by the event args

I'm having trouble understanding the scenario here, could you please help me out, what is it that we should be mocking?
processEventHanlder and processErrorHandler?


>

ACHYUT001 commented on sdk/eventhub/Azure.Messaging.EventHubs/tests/Snippets/Sample10_MockingReadingEventsTests.cs:
```diff
@@ -0,0 +1,254 @@
+                    Assert.AreEqual(lastOffset, properties.Offset);
+                    Assert.AreEqual(fakeDate, properties.EnqueuedTime);
+                    Assert.AreEqual(fakeDate, properties.LastReceivedTime);
+                }
```
Should we assert the values here?

jsquire replied:
We should probably do a "call the metadata method,  assert the values"  pattern,  interleaving the calls.   Something like:

```text
- Get Event Hub Properties
- Assert Event Hub Properties

- Get Partition Properties
- Assert Partition Properties
```

jsquire replied:
On second thought, we can probably just show the query and then leave a comment for "the stuff you just read matches the stuff that you mocked"...

ACHYUT001 replied:
removed the asserts and added comments

ACHYUT001 replied:
please let me know if we should change the language of the comments :)

jsquire replied:
I think the high level of them is great.   We'll eventually want to go with proper capitalization and punctuation to match the Event Hubs codebase.   Here, too, we'll need to double-check that our line lengths fit into the GitHub width to avoid scrolling.   

_(that's easier to do when looking at the markdown, so I suggest we leave it until that point)_

ACHYUT001 commented on sdk/eventhub/Azure.Messaging.EventHubs/tests/Snippets/Sample09_MockingEventHubPublisherTests.cs:
```diff
@@ -0,0 +1,85 @@
+                {
+                    if (!eventBatch.TryAdd(eventData))
+                    {
+                        // My callback rejected the event
```
I see what you mean, we change the callback to return false after it has accepted 2 values.
But in the rejected case, how should we handle the rejection? 
Presently, we are throwing an exception. => we could skip throwing the exception and assert at the end that the values from 3rd position (in the generated expected  events is not present).

ACHYUT001 replied:
we could also have an different test case to demonstrate the rejected scenario 💭

jsquire replied:
hrmm....  I think either approach works, but I like your idea of splitting them out.   I think you're right in that it does help to simplify things.

Commit:
added await operation
refactored code for metadata demos

Commit:
added demonstration for handlers

Commit:
added test for try add false

ACHYUT001 commented:
@jsquire,
I have committed the discussed changes, please review and provide your thoughts :)
Thanks

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/tests/Snippets/Sample10_MockingReadingEventsTests.cs:
```diff
@@ -0,0 +1,290 @@
+                .ReturnsAsync(new string[1] { "sample id" });
+
+            var mockEventHubPartitionReceiver = new Mock<PartitionReceiver>();
+            var mockBatchData = GenerateBatchData();
```
We should consider showing this inline.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/tests/Snippets/Sample10_MockingReadingEventsTests.cs:
```diff
@@ -0,0 +1,290 @@
+
+            string firstPartition;
+
+            await using (var producer = mockEventHubProducerClient.Object)
```
I think we can bypass this part, since we're mocking the receiver directly.   Since we control the partition that we're accepting, then we don't need to show the flow of querying one.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/tests/Snippets/Sample10_MockingReadingEventsTests.cs:
```diff
@@ -0,0 +1,290 @@
+                        cancellationSource.Token);
+
+                    //do we need assertion like this?
+                    for (var index = 0; index < mockBatchData.Count(); ++index)
```
We probably don't need the assert; a comment would cover it well enough, I think.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/tests/Snippets/Sample10_MockingReadingEventsTests.cs:
```diff
@@ -0,0 +1,290 @@
+        private async IAsyncEnumerable<PartitionEvent> GetMockEventsFromPartition(string partitionId,LastEnqueuedEventProperties lastEnqueuedEventProperties = default)
+        {
+            // we are using this as we dont have an async operation in our test
+            await Task.Yield();
```
Something like this would make a good stand-alone sample snippet for "how to mock the the async enumerable with last enqueued properties";  we should consider using a simplified form inline with the "read events" snippets, and having this be it's own

ACHYUT001 commented on sdk/eventhub/Azure.Messaging.EventHubs/tests/Snippets/Sample10_MockingReadingEventsTests.cs:
```diff
@@ -0,0 +1,290 @@
+                .ReturnsAsync(new string[] { mockPartitionId });
+            mockEventHubConsumerClient
+                .Setup(c => c.ReadEventsFromPartitionAsync(It.IsAny<string>(), It.IsAny<EventPosition>(), It.IsAny<CancellationToken>()))
+                .Returns(GetMockEventsFromPartition(mockPartitionId));
```
I see what you mean, it'd be better to keep it to a minimum, so that it's easier to understand.
I could always write an overload to the `GetMockEvents()` with last enqueued event and use that for a new sample.

ACHYUT001 replied:
However, I dont see the `EventData` constructor which accepts string:
![image](https://user-images.githubusercontent.com/25906933/124164941-e674fc80-dabe-11eb-86b2-5bfe08c5eb8c.png)

ACHYUT001 replied:
![image](https://user-images.githubusercontent.com/25906933/124165041-07d5e880-dabf-11eb-88aa-62a29f519ad7.png)

jsquire replied:
Rebase on `main`, it's a recent commit.  😄

jsquire replied:
https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs/src/EventData.cs#L419

ACHYUT001 replied:
updated the function to generate `EventData` inline

jsquire replied:
Cool, it looks good!   Now we just need to hoist it into the snippet as a local function so that everything is available to the reader at a glance.

Commit:
refactored code
removed unnecessary mocks and assert statements
added batch event generation inline

ACHYUT001 commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Snippets/Sample07_MockingEventProcessorHandlersTest.cs:
```diff
@@ -0,0 +1,114 @@
+{
+    [TestFixture]
+    public class Sample07_MockingEventProcessorHandlersTest
+    {
```
Have separated the Event and Error Handler code

jsquire replied:
This looks great!   It really cut down to just the focus area.

ACHYUT001 commented on sdk/eventhub/Azure.Messaging.EventHubs/tests/Snippets/Sample10_MockingReadingEventsTests.cs:
```diff
@@ -0,0 +1,292 @@
+        ///
+        [Test]
+        public async Task MockingReadEventsFromPartition_WithLastEnqueuedEventProperties()
+        {
```
added a new test method to demonstrate reading events with last enqueued event properties, this uses the `GetsMockEventFromPartition` overload with `lastEnqueuedEventPartitions`

ACHYUT001 replied:
let me know if we need this test method or just the local `GetsMockEventFromPartition` overload with `lastEnqueuedEventPartitions` is enough.

jsquire replied:
If we make a local function out of `GetMockEventsFromPartition` and demonstrate calling it with the last enqueued mock, I think that's a good way to focus on what we're trying to highlight here.   We can probably kill the consumer bits, call `GetMockEventsFromPartition`  and demonstrate using the partition context to return the mocked data.   Maybe we consider a trailing comment on being able to use this pattern while mocking out consumers.

What are your thoughts?

ACHYUT001 replied:
just reiterating/ thinking out loud here : 
we should not mock the consumer, and call the `GetMockEventsFromPartition()` directly and use that to return the mocked data using the returned partition context?

What scenario's would one want to not use this pattern?
Is the idea to demonstrate using `PartitionContext` to fetch mock data w/o using the `ConsumerClient`?

jsquire replied:
We want each snippet to be as self-contained as possible; ideally, a developer should be able to copy/paste the sample and it should compile/run without having to hunt down common utility methods in other snippets.  For clarity, it's better to demonstrate a concept in isolation than to introduce complexity with trying to demonstrate too many things at once.

If we feel that showing the generation doesn't work without the consumer bits, I'm good with that.   We'll want to break things down into multiple snippets that we can render back-toback since there's a clear functionality delineation, which also potentially allows us to reuse the same utility method snippet elsewhere.  For example:

```csharp
#region Snippet:EventHubs_Sample10_GeneratingMockEvents
public async IAsyncEnumerable<PartitionEvent> GetMockEventsFromPartition(int partitionId)
{   
    var partitionContext = EventHubsModelFactory.PartitionContext(partitionId,);

    await Task.Yield();
    yield return new PartitionEvent(partitionContext, new EventData("One"));
    yield return new PartitionEvent(partitionContext, new EventData("Two"));
    yield return new PartitionEvent(partitionContext, new EventData("Three"));
}
#endregion
```
```csharp
#region Snippet:EventHubs_Sample10_UsingMockEventsWithTheConsumer
// .. THE CONSUMER BITS
#endregion
```

For a more complex illustration, we break things down in a couple of different places, in different ways:

- In our basic [Hello World](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample01_HelloWorld.md) sample, we introduce creating clients in their own snippet and mention that we're assuming this pattern below for the more complex cases.   This works because it's simple and foundational.

- In the [Migration Guide](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs/MigrationGuide.md#migrating-eventprocessorhost-checkpoints), we demonstrate moving checkpoints in a few snippets that appear together in a section, where each builds on the other.   We do this to try and reduce complexity by talking over independent parts independently.  Here, things are tightly clustered because they are somewhat complex and their usage is really isolated to this particular scenario.

ACHYUT001 commented on sdk/eventhub/Azure.Messaging.EventHubs/tests/Snippets/Sample10_MockingReadingEventsTests.cs:
```diff
@@ -0,0 +1,292 @@
+                    cancellationSource.Token))
+                {
+                    string readFromPartition = partitionEvent.Partition.PartitionId;
+                    ReadOnlyMemory<byte> eventBodyBytes = partitionEvent.Data.EventBody.ToMemory();
```
just a thought do we need to add an assert or comment here, to point out the mocked data is being returned?

jsquire replied:
Good call.   I'm comfortable with what you decide;  if it were me, I'd probably go the lazy route and just add a comment.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Snippets/Sample07_MockingEventProcessorHandlersTest.cs:
```diff
@@ -0,0 +1,114 @@
+            var mockEventBody = new BinaryData("This is a sample event body");
+            var mockEventData = EventHubsModelFactory.EventData(mockEventBody);
+
+            var mockProcessEventArgs = new ProcessEventArgs(mockPartitionContext, mockEventData, updateToken => Task.CompletedTask, CancellationToken.None);
```
Once we get the overall form to the point that we're comfortable, we'll want to keep an eye on line length;  the samples get read on GitHub more often than downloaded.  We've had some strong feedback to earlier samples that we needed to make that we fit into the GitHub width to keep things readable.

ACHYUT001 replied:
sounds reasonable, so I should probably push updateToken to next line then...

andreyshihov commented:
> How do I mock invoking my event and error handlers for the `EventProcessorClient`? _(separate sample in the processor package)_
> 
> * Demonstrate using the `EventHubsModelFactory` to create the types needed by the event args

Hi @jsquire, I have just created draft PR #23677 to demonstrate use of Factory in the Handler tests. **Am I moving in the right direction?** Please share your thoughts. This is just first steps to understand if the concept does makes sense.

If concept is good, I'm thinking about adding more tests demonstrating how to verify Handler's behavior with following mockable properties:
//eventArgs.Data.EventBody; 
//eventArgs.Data.PartitionKey; 
//eventArgs.Data.Properties; 
//eventArgs.Data.Offset; 
//eventArgs.Data.SequenceNumber; 
//eventArgs.Data.SystemProperties; 

Does it makes sense?

jsquire commented:
Hi @ACHYUT001. Thanks very much for your contribution and helping to improve the Azure SDK experience.  Since this work stream has been inactive for a while and another community member has expressed interest in taking this on,  I'm going to close this out.   

I appreciate your efforts and would welcome the opportunity to collaborate again in the future should you wish to do so.
